Specify the minimum JWE server requirements

sberyozkin opened the pull request:
Fixes #58. 

@rdebusscher, @starksm64, All,
This document update introduces the minimum JWE server requirements and I've tried to make sure it is all as simple and unambiguous as it can get. Only a single new property is introduced which is sufficient to support both the claims encrypted directly and the encrypted nested tokens.

The implementation flow becomes: If `mp.jwt.decrypt.key.location` is set then it must be the encrypted content otherwise - signed as usual. When dealing with the encrypted content, once the decryption is done, the implementation checks if the verification properties are also set - if yes - the decrypted content is a signed JWT token and must go through the current verification step, otherwise these are claims ready to be converted to `JsonWebToken`. 

This is really it, pretty basic, and IMHO it should not become any more sophisticated in the short to medium term. `RSA-OAEP` and `A256GCM` algorithms are the only ones which are required and is in sync with the JWT Builder API PR.

Note IMHO `mp.jwt.decrypt.key.location` is better than `mp.jwt.decrypt.privatekey.location`, as even though no `dir` or key wrapping which would require the use of the symmetric keys are supported, I'd rather avoid having to introduce one more location property specifically to support the symmetric keys when such requirements will occur.

I'm going to work next on a few TCK tests. In meantime please provide some feedback, I'd particularly appreciate some advice about having a better English text in this PR :-)

sberyozkin commented:
@rdebusscher I've updated `TokenUtils` in preparation to starting adding few TCK tests:
- deprecated `generateTokenString` methods and added the equivalent `signClaims` methods
- added the same number of `encryptClaims` methods
- and two methods for now to have the claims signed and then encrypted. `encryptSignedClaims` as the actual process of signing and encryption is already covered

sberyozkin commented:
While it would be straightforward to supper either signed or encrypted or whatever tokens (just check if it is 3 vs 5 dots and may be also check the media type), it is too flexible for the practical deployments - they would usually decide what the security policy is, as opposed to: you can either sign or encrypt or sign and encrypt - it does not really that dynamic. This can be reviewed going forward, but to start with it would be better IMHO just to require a single type of protection at the MP JWT level.

sberyozkin commented:
@rdebusscher @starksm64, others, This is now ready for review.

I've verified the 2 TCK tests I've added against my branch, and will add few more tests (to enforce the text related to how the configuration properties determine what sort of the token protection is expected).

In its essence it is a small but important enhancement. As I said, I've tried my best to make it unambiguous and also straightforward to implement (trust me, it was pretty easy to implement :-) ).

Lets try to get it into 1.2. Thanks

sberyozkin commented:
I'd also to thank Roberto for his work with the GitHub actions and TCK related work in our project which helped me to accelerate

sberyozkin commented:
@rdebusscher @starksm64 I've added few configurations tests as well, and consider this PR quite complete now. I might add some more TCKs but I'd like to avoid creating temp files, etc. The spec text is covered reasonably well with the TCK now.

sberyozkin commented on spec/src/main/asciidoc/configuration.asciidoc:
```diff
@@ -352,13 +360,71 @@ See https://docs.oracle.com/javase/8/docs/api/java/net/URL.html[java.net.URL] ja
+
+The rules about matching a token `kid` header and JWK `kid` property for selecting the verification keys apply when a key management key is formatted as JWK.
+
+If the claims have been immediately encrypted, without being signed first, then the application endpoints will have no guarantee that a token came from a trusted issuer. To have this proof the claims will need to be signed first and the resulting nested JWT token - encryped next.
```
@rdebusscher - now that we support the cookies, we should expect that the users will actually encrypt only (hopefully with the help of the builder API :-) in 2.0) when sign-encrypt would be redundant/unnecessarily expensive. 
So IMHO this requirement should stay if the token comes with a cookie. And make it optional if it comes with `Authorization`. This is an easy way for the implementations to choose not to support it if it is `Authorization` while reacting as needed to the concrete requirements we may not be aware of right now, one example: 
Mutual TLS is used, the clients sends an encrypted token, expecting the MP JWT server propagating it further. MTLS gives the assurance the token came from the trusted source, while the encryption alone gives an option to retain the confidentiality outside of the HTTPS. 

Let me update the text

rdebusscher replied:
Cookie usage would be a very exceptional case as Cookies are used for Web applications, not for REST endpoints.

but
>I'm making the encrypted-only totally optional, and only recommend accepting them if they are coming as cookies

is a good compromise.

And security has nothing to do with performance, in the contrary, so

>redundant/unnecessarily expensive.

is a dangerous statement.

sberyozkin replied:
Hi @rdebusscher, I think we can agree that we all appreciate here the importance of a non-compromise good security. My point has been very simple, when someone creates a cookie, self-signing it and then encrypting is IMHO not adding anything as compared to a plain encryption, so in this specific case I'll stand by my statement above :-)

sberyozkin commented on spec/src/main/asciidoc/configuration.asciidoc:
```diff
@@ -352,13 +360,71 @@ See https://docs.oracle.com/javase/8/docs/api/java/net/URL.html[java.net.URL] ja
+In such cases, when the confidentiality of the claims is critical, the claims can be encrypted
+(instead of being signed), producing a JWT token by following the steps in the https://tools.ietf.org/html/rfc7516[JSON Web Encryption(JWE)] specification.
+
+Decrypting the encrypted claims requires a single property, `mp.jwt.decrypt.key.location`, pointing to a private key which can be used to decrypt the token. All the location options supported by the `mp.jwt.verify.publickey.location` property are also supported by `mp.jwt.decrypt.key.location`.
```
`mp.jwt.verify.issuer` is at the lower level :-), it applies irrespectively of how the claims have been secured, signed/encrypted/signed-and-encrypted. I've actually updated the text there to note about it.
I know that the verification keys are also in the `.verify.' space, same as the issuer, and soon, - audience, but it is fine, the text clarification that the verification of the claims goes next after the claims protection has been asserted (signature validation, correct decryption) is enough IMHO

sberyozkin replied:
@rdebusscher I'm adding the extra clarification under the actual property definition. I think we can create a dedicated Claims Verification section in time to stress a difference between verifying the token integrity/decryption and the actual content verification, but probably not right now

sberyozkin commented on spec/src/main/asciidoc/configuration.asciidoc:
```diff
@@ -352,13 +360,71 @@ See https://docs.oracle.com/javase/8/docs/api/java/net/URL.html[java.net.URL] ja
+An extra care is required to ensure the private keys are not leaked, particularly, making them available at the insecure HTTP locations or as part of the application archive is not recommended. This is also why the inlined private keys are not supported.
+
+Note that two types of keys are required to encrypt the claims:
+* Content encryption key - typically a secret key which is generated during the encryption process and used to encrypt a sensitive content such as a JSON representation of the token claims.
```
@rdebusscher I've tried to keep it to the minimum. I think we should expect the developers have some high level understanding of what is involved.

sberyozkin commented on spec/src/main/asciidoc/configuration.asciidoc:
```diff
@@ -352,13 +360,71 @@ See https://docs.oracle.com/javase/8/docs/api/java/net/URL.html[java.net.URL] ja
+* Content encryption key - typically a secret key which is generated during the encryption process and used to encrypt a sensitive content such as a JSON representation of the token claims.
+* Key management key - public RSA key which is used to encrypt a content encryption key. `mp.jwt.decrypt.key.location` must point to a private RSA key matching this key.
+
+Key management key algorithm which must be supported is https://tools.ietf.org/html/rfc7518#section-4.3[RSA-OAEP] (RSAES using Optimal Asymmetric Encryption Padding) with a key length 2048 bits or higher.
```
@rdebusscher I'm sure some texts refer to it as such, but we are working against the [JWA spec](https://tools.ietf.org/html/rfc7518#section-4) which qualifies it as `Key Management` - which implies the encryption of the content encryption keys. 

`Encryption scheme`, according to this [wikipedia text](https://en.wikipedia.org/wiki/Encryption), actually describes the encryption of the plaintext. So I may be able to update the `Content Encryption Key` text a bit, something like:
```
Content encryption key - it supports an encryption scheme where a secret key which is generated during the encryption process is used to encrypt a sensitive content such as a JSON representation of the token claims (plaintext).
```
That will match quite closely the wikipedia text :-)

sberyozkin replied:
@rdebusscher I think I may have simplified a bit this short intro and also injected your `encryption scheme` term

sberyozkin commented:
@rdebusscher Thanks for the quality review as usual, I'll work on the comments

sberyozkin commented:
@rdebusscher I believe I've addressed your requests, please do another round.

sberyozkin commented:
@rdebusscher Actually, let me create a dedicated section called `Verification of the token claims` and move the issuer and also the audiences there, that would be definitely clearer

sberyozkin commented:
@rdebusscher I'm actually quite happy now with the updated text as there is a clear enough structure around it with the cross-references:
- Signed Jwt Tokens
  - Signature Verification Configuration Parameters
- Encrypted Jwt Tokens and nested tokens
  - Decryption Configuration Parameters
- Verification of JWT token claims
  (this is linked to from both signed and encrypted token sections)
- Rules for accepting signed and encrypted tokens
- JWT and HTTP headers
- Rules for MP JWT Configuration Parameters
  - mappings to env

It might take a couple of more rounds to tune it even better but IMHO it is not too bad now

sberyozkin commented:
@rdebusscher Hi, I'd like to go ahead with merging this PR asap, do you have any other concerns ? Please do another review round.

rdebusscher commented on tck/src/test/java/org/eclipse/microprofile/jwt/tck/container/jaxrs/jwe/RolesAllowedEncryptTest.java:
```diff
@@ -0,0 +1,300 @@
+
+    @RunAsClient
+    @Test(groups = TEST_GROUP_JAXRS, description = "Validate a request with no token fails with HTTP_UNAUTHORIZED")
+    public void callEchoNoAuth() throws Exception {
```
Exception not thrown in method so can be removed.

This is the case for most test methods in the class.

rdebusscher commented on tck/src/main/java/org/eclipse/microprofile/jwt/tck/util/TokenUtils.java:
```diff
@@ -90,31 +123,287 @@ public static String generateTokenString(final String jsonResName, final Set<Inv
+     * @return the JWT string
+     * @throws Exception on parse failure
+     */
+    public static String signClaims(PrivateKey pk, String jsonResName) throws Exception {
```
Not used

rdebusscher commented on tck/src/test/java/org/eclipse/microprofile/jwt/tck/util/TokenUtilsTest.java:
```diff
@@ -42,26 +42,52 @@
+    }
+    
+    @Test(groups = TCKConstants.TEST_GROUP_UTILS, expectedExceptions = {InvalidJwtException.class},
+            description = "Deprecated: Illustrate validation of alg")
```
Why are these just deprecated and not just removed.

Deprecated doesn't make sense for TCK IMHO.

sberyozkin replied:
@rdebusscher I've seen `TokenUtils` being used outside of TCK, certainly our additional product tests refer to it as far as I recall. It is very feasible someone else, even the users, try them since it is the only support currently available around MP JWT itself to generate the token.

sberyozkin commented on spec/src/main/asciidoc/configuration.asciidoc:
```diff
@@ -352,15 +341,81 @@ See https://docs.oracle.com/javase/8/docs/api/java/net/URL.html[java.net.URL] ja
+
+## Requirements for accepting signed and encrypted tokens
+
+This specification does not support the application endpoints accepting signed or encrypted or signed and encrypted tokens at the same time. Requiring the implementations to determine a token type at runtime (signed token will have 3 parts, encrypted token - 5 parts) will have a small performance cost and it is expected that a given MP JWT application will accept only signed or only encrypted or  ...
```
@rdebusscher I agree it is technically easy to implement, but I'd like the spec to start from the assumption that the the endpoint is not accepting all types of tokens at the same time - because the policy is typically determined in advance - what kind of protection is expected around a token. 
If my requirement is for the application to accept signed then encrypted tokens then if we relax as you suggest and someone sends encrypted-only or signed - only then the application will still accept the tokens despite the requirements being more stringent.

I propose that we keep this restriction for now and if there will be some practical demand we ill easily relax it, we can introduce a property which will users to avoid ambiguities in cases like the one described above (`mp.jwt.token.security=all` or something like that).

Hope it sounds reasonable

sberyozkin commented:
Hi @rdebusscher I've worked on addressing your comments, resolved few obvious comments.
Re the dynamic token type support, so yes, as proposed, IMHO it would be better to start with a more restrictive approach, and indeed we'll relax it in no time should some users offer some sound practical cases. In general JOSE does not really `like` the dynamism though but we can handle it if needed with the extra properties.
But I've added an extra check around `cty` - it was already set in the `TokenUtils` but was not tested and TCK also now enforces it is set in the `sign-encrypt` case. This is a little bit inconsistent with the fact that `typ` is not enforced for the directly secured claims, but that can be discussed in a separate PR I'll open later to do with clarifying what the required claims are (I don't think we can enforce `typ` it in `1.2` or even should enforce it). But requiring `cty` in the `sign-encrypt` case is reasonable, I don't think OIDC servers can do `sign-encrypt` so in a way we can drive this requirement early in MP JWT; the builder API PR docs `cty` has to be set automatically in the `innerSign().encrypt()` flow.

So all in all this PR now looks very close to getting merged IMHO, please do another check and lets merge it asap
thanks

sberyozkin commented:
@rdebusscher Hi Rudy, I'm getting ready to merge it as I believe I've addressed your comments - I will hold off though till Fri in case you or other colleagues will have some more concerns. I can assure everyone every concern will be addressed
Thanks

sberyozkin commented:
@rdebusscher Please create a new change request now or later, I don't want to rush things, but it would be easier for me to start the new items. I think we've had a good discussion so far, cheers

sberyozkin commented:
Hi @rdebusscher I've been thinking again if we can make it explicitly optional to support all token types at the same time, but after a long deliberation it still feels like we'd be making a step backward as to avoid the ambiguity described above, when it is needed, the implementations will need to start coming up with their own properties etc so IMHO it really makes sense at this stage to be a bit more conservative about it, as I mentioned the dynamism (where the client dictates the format or the alg etc) is not usually very welcome
thanks

sberyozkin commented:
@rdebusscher OK, I have to commit to continue building on top of it. Thanks a mill for the review.
All, any concerns/suggestions about this feature - please open an issue
